packages/xr/src/lib/components/teleport-controls.svelte

```diff
@@ -0,0 +1,179 @@
+
+@param handedness - Which hands to allow teleportation from.
+
+@param maxDistance - The maximum radial teleporting distance from the user's current origin, in meters.
+
+@event teleport - Fires after a teleport event.
+
+```svelte
+  <TeleportControls
+    handedness={'left' | 'right'}
+    maxDistance={10}
+    on:teleport={(event) => {}} 
+  >
+    <T.Mesh teleportSurface>
+      ...
+    </T.Mesh>
+  </TeleportControls>
+```
+-->
+
+<script lang='ts'>
+
+import * as THREE from 'three'
+import { T, useFrame, createRawEventDispatcher } from '@threlte/core'
+import { activeTeleportController, pendingTeleportDestination } from '../internal/stores'
+import { useTeleport, useController, useGamepad } from '../hooks'
+import Ray from '../components/ray.svelte'
+import { navMeshes, teleportPlugin } from '../plugins/teleport-controls'
+
+/**
+ * The raycaster used for teleportation.
+ * @default new THREE.Raycaster()
+ */
+export let raycaster = new THREE.Raycaster()
+
+/**
+ * The controller handedness that the teleport controls is linked to.
+ * @default 'right'
+ */
+export let handedness: 'left' | 'right' = 'right'
+
+/**
+ * The maximum radial teleporting distance from the user's current origin, in meters.
+ * @default 20
+ */
+export let maxDistance = 20
+
+type $$Events = {
+  /** Fired after a teleportation occurs with the new location as the payload. */
+  teleport: THREE.Vector3
+}
+
+const dispatch = createRawEventDispatcher<$$Events>()
+
+let destination: THREE.Vector3 | undefined
+let activeController: THREE.XRTargetRaySpace | undefined
+
+const teleport = useTeleport()
+const controllerPosition = new THREE.Vector3()
+const matrix4 = new THREE.Matrix4()
+const rayMidpoint = new THREE.Vector3()
+const rayDivisions = 40
+const positions = new Float32Array(rayDivisions * 3)
+const curve = new THREE.QuadraticBezierCurve3()
+const curvePoint = new THREE.Vector3()
+
+$: raycaster.far = maxDistance
+$: teleportController = useController(handedness)
+$: teleportGamepad = useGamepad(handedness)
+
+const calculateRayMidpoint = (vector1: THREE.Vector3, vector2: THREE.Vector3) => {
+  rayMidpoint.x = (vector1.x + vector2.x) / 2;
+  rayMidpoint.y = (vector1.y + vector2.y) / 2;
+  rayMidpoint.z = (vector1.z + vector2.z) / 2;
+}
+
+const { start, stop } = useFrame(() => {
+  matrix4.identity().extractRotation(activeController!.matrixWorld)
+  raycaster.ray.origin.setFromMatrixPosition(activeController!.matrixWorld)
+  raycaster.ray.direction.set(0, 0, -1).applyMatrix4(matrix4)
+
+  const [intersect] = raycaster.intersectObjects(navMeshes, true)
+
+  if (intersect === undefined) {
+    destination = undefined
+    $pendingTeleportDestination = undefined
+    return
+  }
+
+  destination = intersect.point
+  $pendingTeleportDestination = destination
+
+  activeController!.getWorldPosition(controllerPosition)
+
+  calculateRayMidpoint(controllerPosition, destination)
+
+  // Create an arc
+  rayMidpoint.y += 0.8
```

Comment:
I think so! I'd like to also make the arc height generally a function of the length and then allow the user to tune some base constant. Right now since it's always the same it's too arc-y when short and not arc-y enough when long.

Opposed to me adding this in a follow up?